File: pkg/legionplugin/tool.go
// Package legionplugin is the Go guest SDK for Legion Agent WASM plugins
// (ABI v1).
//
// A whole plugin looks like this:
//
//	package main
//
//	import "github.com/stardust/legion-agent/pkg/legionplugin"
//
//	func init() {
//		legionplugin.Serve("legion-hello-go", "0.1.0", legionplugin.Tool{
//			Name: "hello_echo",
//			Handler: func(call legionplugin.ToolCall) legionplugin.ToolResult {
//				name := call.Argument("name")
//				if name == "" {
//					return legionplugin.Fail("missing required argument: name")
//				}
//				legionplugin.LogInfo("hello_echo called with name=" + name)
//				return legionplugin.OK("hello, " + name + "!")
//			},
//		})
//	}
//
//	func main() {}
//
// Build it with:
//
//	GOOS=wasip1 GOARCH=wasm go build -buildmode=c-shared -o plugin.wasm .
//
// # Why registration goes in init and main stays empty
//
// A plugin is a WASI *reactor*, not a command: the host instantiates it with
// WithStartFunctions("_initialize") and then calls exported functions. Nothing
// ever calls main. Package initialization — which is what _initialize runs —
// is therefore the only place registration can happen, and main exists solely
// because package main requires it.
//
// # Capabilities
//
// LogInfo is available unconditionally. The other six host functions are
// behind build tags (legion_config, legion_kv, legion_http, legion_fs,
// legion_tool) for the reason spelled out in host_wasip1.go: importing a host
// function the deployment did not grant makes the module fail to INSTANTIATE,
// so "import only what you use" is a hard rule rather than a preference.
package legionplugin

import (
	"encoding/json"
	"errors"
	"fmt"
)

// ToolCall is one call the host hands this plugin.
//
// CallID is here so a plugin can correlate its own logs with the host's; it is
// NOT echoed back. The host owns the correlation id and overwrites whatever a
// guest returns, which is why ToolResult has no such field.
type ToolCall struct {
	CallID string `json:"call_id"`
	Tool   string `json:"tool"`
	// Arguments is string-to-string because that is the host's own type for it
	// (internal/plugin/host.guestToolCall): numbers and booleans arrive as
	// their string spellings and the plugin parses them itself.
	Arguments map[string]string `json:"arguments"`
}

// Argument returns a required argument, or "" when it is missing OR empty.
//
// Treating an empty value as absent is deliberate: an argument that is only
// whitespace is almost always a caller mistake, and a result computed from one
// is a result nobody can explain.
func (c ToolCall) Argument(name string) string {
	return c.Arguments[name]
}

// ToolResult is what a plugin answers, mirroring the host's domain.ToolResult.
//
// Report failure with Fail, never with a panic: a panic traps the whole wasm
// module — costing the instance's state, every other call in flight on it, and
// a fault against the plugin's health — while a failed result is just an
// answer the model can read and react to.
type ToolResult struct {
	success bool
	output  string
	failure string
}

// OK builds a successful result. output is what the model sees.
func OK(output string) ToolResult {
	return ToolResult{success: true, output: output}
}

// Fail builds a failed result. message should say what was wrong; it reaches
// the model verbatim.
func Fail(message string) ToolResult {
	return ToolResult{success: false, failure: message}
}

// wireResult is the document the host strictly decodes. Every field is written
// out, none omitted: the host's domain.ToolResult has no omitempty, so leaving
// fields out would only give the two sides different readings of one document.
type wireResult struct {
	CallID  string `json:"call_id"`
	Success bool   `json:"success"`
	Output  string `json:"output"`
	Error   string `json:"error"`
}

// encode renders the result as the host's ToolResult JSON.
//
// call_id is always empty: the host overwrites it, so filling it in here would
// only mislead whoever reads the guest's own logs.
func (r ToolResult) encode() ([]byte, error) {
	body, err := json.Marshal(wireResult{
		Success: r.success,
		Output:  r.output,
		Error:   r.failure,
	})
	if err != nil {
		return nil, fmt.Errorf("encode tool result: %w", err)
	}
	return body, nil
}

// Handler implements one tool.
type Handler func(ToolCall) ToolResult

// Tool is one tool this plugin contributes: the name the deployment's
// plugin.json must also list, and the function behind it.
type Tool struct {
	Name    string
	Handler Handler
}

// parseToolCall decodes an abi.OpCallTool request body.
//
// A missing tool name is an error rather than an empty name: an empty name
// would be dispatched as "unknown tool", which blames the caller for the SDK's
// own silence. Unknown fields are NOT refused — the host may add an optional
// field later, and a plugin that failed on it would be a fault the deployment
// counts against its health.
func parseToolCall(body []byte) (ToolCall, error) {
	var call ToolCall
	if err := json.Unmarshal(body, &call); err != nil {
		return ToolCall{}, fmt.Errorf("parse tool call: %w", err)
	}
	if call.Tool == "" {
		return ToolCall{}, errors.New("parse tool call: no tool name")
	}
	return call, nil
}

// manifestDoc is op 0's self-description. Provides is derived from the
// registered tools, so what the guest says it implements and what it actually
// dispatches cannot drift apart — a mismatch is what makes activation's
// cross-check refuse to mount the package.
type manifestDoc struct {
	Name     string   `json:"name"`
	Version  string   `json:"version"`
	Provides []string `json:"provides"`
}

File: sdk/rust/legion-plugin/src/tool.rs
//! 工具调用与结果：插件作者唯一需要读懂的两个类型。

use crate::json;
use std::collections::BTreeMap;

/// ToolCall 是宿主交给插件的一次调用。
///
/// `call_id` 只用于把插件自己的日志和宿主的日志对上——**答回去时不要填它**，
/// 相关性 id 归宿主所有，它会覆盖插件返回的值（`ToolResult` 因此也没有这个字段）。
pub struct ToolCall {
    pub call_id: String,
    pub tool: String,
    /// arguments 是字符串到字符串的映射：宿主侧的类型就是 `map[string]string`，
    /// 所以数字、布尔在这里都是它们的字符串写法，由插件自己解析。
    pub arguments: BTreeMap<String, String>,
}

impl ToolCall {
    /// parse 读一次 `abi::OP_CALL_TOOL` 的请求体。
    ///
    /// 缺 `tool` 是错误而不是空名字：一个空工具名会被分发当成「未知工具」，那
    /// 句错误信息会把 SDK 的沉默说成调用方的错。
    pub fn parse(body: &[u8]) -> Result<ToolCall, json::ParseError> {
        let object = json::parse(body)?;
        let tool = object
            .strings
            .get("tool")
            .filter(|value| !value.is_empty())
            .ok_or(json::ParseError::MissingField("tool"))?
            .clone();
        Ok(ToolCall {
            call_id: object.strings.get("call_id").cloned().unwrap_or_default(),
            tool,
            arguments: object.arguments,
        })
    }

    /// argument 取一个必填参数，缺失或为空时返回 `None`。
    ///
    /// 空串与缺失同等对待是刻意的：一个只有空白的参数几乎总是调用方写错了，而
    /// 拿它继续跑出来的结果谁也解释不了。
    pub fn argument(&self, name: &str) -> Option<&str> {
        self.arguments
            .get(name)
            .map(String::as_str)
            .filter(|value| !value.is_empty())
    }
}

/// ToolResult 是插件答给宿主的东西，对应宿主侧的 `domain.ToolResult`。
///
/// **失败要用 `ToolResult::fail`，不要 panic**：panic 会 trap 整个 wasm 模块，
/// 代价是实例里的状态和同一实例上所有在途调用，而且会被记进插件的健康度；失败
/// 结果只是模型能读懂、能据此改口再试的一个答案。
pub struct ToolResult {
    success: bool,
    output: String,
    error: String,
}

impl ToolResult {
    /// ok 构造一个成功结果，output 是给模型看的内容。
    pub fn ok(output: impl Into<String>) -> ToolResult {
        ToolResult {
            success: true,
            output: output.into(),
            error: String::new(),
        }
    }

    /// fail 构造一个失败结果。message 要说清是什么不对——它会原样到达模型。
    pub fn fail(message: impl Into<String>) -> ToolResult {
        ToolResult {
            success: false,
            output: String::new(),
            error: message.into(),
        }
    }

    /// to_json 渲染宿主严格解码的那份文档。
    ///
    /// `call_id` 恒为空串：宿主拥有相关性 id 并会覆盖它（见 ToolCall::call_id）。
    /// 四个字段全部写出而不省略，因为宿主侧的 `domain.ToolResult` 没有
    /// `omitempty`，省略字段只会让两边对同一份文档的读法产生分歧。
    pub fn to_json(&self) -> String {
        format!(
            "{{\"call_id\":\"\",\"success\":{},\"output\":\"{}\",\"error\":\"{}\"}}",
            self.success,
            json::escape(&self.output),
            json::escape(&self.error)
        )
    }
}

/// ToolHandler 是一个工具的实现：拿到调用，给出结果。
pub type ToolHandler = fn(&ToolCall) -> ToolResult;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_a_tool_call_with_arguments() {
        let call = ToolCall::parse(br#"{"call_id":"c1","tool":"hello","arguments":{"name":"legion"}}"#)
            .expect("a well-formed call must parse");
        assert_eq!(call.call_id, "c1");
        assert_eq!(call.tool, "hello");
        assert_eq!(call.argument("name"), Some("legion"));
    }

    #[test]
    fn a_missing_tool_field_is_an_error_not_an_empty_name() {
        assert_eq!(
            ToolCall::parse(br#"{"call_id":"c1","arguments":{}}"#).err(),
            Some(json::ParseError::MissingField("tool"))
        );
    }

    #[test]
    fn an_empty_argument_reads_as_absent() {
        let call = ToolCall::parse(br#"{"tool":"hello","arguments":{"name":""}}"#).expect("parse");
        assert_eq!(call.argument("name"), None);
    }

    #[test]
    fn ignores_fields_this_sdk_does_not_consume() {
        // 宿主日后加一个可选字段，不该把每个插件都变成故障。
        let call = ToolCall::parse(br#"{"tool":"hello","deadline_ms":3000,"arguments":{"a":"b"}}"#)
            .expect("an unknown scalar field must not fail the parse");
        assert_eq!(call.tool, "hello");
        assert_eq!(call.argument("a"), Some("b"));
    }

    #[test]
    fn escapes_quotes_and_backslashes_in_results() {
        let json = ToolResult::ok("say \"hi\"\\ok").to_json();
        assert!(json.contains("say \\\"hi\\\"\\\\ok"), "got {json}");
    }

    #[test]
    fn escapes_newlines_so_the_host_can_decode_the_body() {
        let json = ToolResult::ok("line one\nline two").to_json();
        assert!(json.contains("line one\\nline two"), "got {json}");
        assert!(!json.contains('\n'), "a raw newline would break the host's strict decode: {json}");
    }

    #[test]
    fn a_failed_result_carries_the_message_and_success_false() {
        let json = ToolResult::fail("missing required argument: name").to_json();
        assert!(json.contains("\"success\":false"), "got {json}");
        assert!(json.contains("missing required argument: name"), "got {json}");
    }
}
